x/pagination.go

```diff
@@ -18,3 +21,42 @@ func ParsePagination(r *http.Request) (page, itemsPerPage int) {
 func PaginationHeader(w http.ResponseWriter, u *url.URL, total int64, page, itemsPerPage int) {
 	migrationpagination.PaginationHeader(w, u, total, page, itemsPerPage)
 }
+
+type Page struct {
+	Page, ItemsPerPage int
+}
+
+var PagePaginationLimit = 1000
+
+func ParseKeysetOrPagePagination(r *http.Request) ([]keysetpagination.Option, *Page, error) {
+	q := r.URL.Query()
+	// If we have any new-style pagination parameters, use those and ignore the rest.
+	if q.Has("page_token") || q.Has("page_size") {
+		keyset, err := keysetpagination.Parse(q, keysetpagination.NewStringPageToken)
+		if err != nil {
+			return nil, nil, herodot.ErrBadRequest.WithReason(err.Error())
+		}
+		return keyset, nil, nil
+	}
+	// allow fallback page pagination with upper limit
+	if q.Has("page") {
+		paginator := pagepagination.PagePaginator{MaxItems: 500, DefaultItems: 250}
+		page, perPage := paginator.ParsePagination(r)
+		if page*perPage > PagePaginationLimit {
+			return nil, nil, herodot.ErrBadRequest.WithReasonf("Legacy pagination is not supported for enumerating over %d items. Please switch to using page_token and page_size.", PagePaginationLimit)
+		}
+		return nil, &Page{page, perPage}, nil
+	}
+	// Allow passing per_page instead of page_size if only the former is set...
+	if q.Has("per_page") && !q.Has("page_size") {
+		q.Set("page_size", q.Get("per_page"))
+		q.Del("per_page")
+		r.URL.RawQuery = q.Encode()
+	}
+	// ... and defaul to keyset pagination
+	keyset, err := keysetpagination.Parse(q, keysetpagination.NewStringPageToken)
+	if err != nil {
+		return nil, nil, herodot.ErrBadRequest.WithReason(err.Error())
+	}
+	return keyset, nil, nil
+}
```

Comment:
Should we move this to ory/x instead?

Reply:
I'm back and forth on this. We have 5 pagination packages in ory/x already, and they are still not complete.

I _would_ like to improve and clean up the pagination packages, but that's for another PR.